File: 0303-Range-Sum-Query-Immutable/0303.py
from typing import List

class NumArray:

    # nums:       [1,2,-3,0,4]
    # prefix_sum: [1,3,0,0,4]
    # sumRange = prefix_sum[right] - prefix_sum[left - 1]
    
    def __init__(self, nums: List[int]):
        self.prefix_sum = []
        sum = 0
        for num in nums:
            sum += num
            self.prefix_sum.append(sum)

    def sumRange(self, left: int, right: int) -> int:
        right = self.prefix_sum[right]
        left = self.prefix_sum[left - 1] if left > 0 else 0
        return  right - left



# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# param_1 = obj.sumRange(left,right)

File: 0303-Range-Sum-Query-Immutable/0303.java
class NumArray {

    private List<Integer> prefixSum = new ArrayList<>();
    private int sum = 0;

    public NumArray(int[] nums) {
        for (int num : nums) {
            sum += num;
            prefixSum.add(sum);
        }
    }

    public int sumRange(int left, int right) {
        int R = prefixSum.get(right);
        int L = left > 0 ? prefixSum.get(left - 1) : 0;
        return R - L;
    }
}

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray obj = new NumArray(nums);
 * int param_1 = obj.sumRange(left,right);
 */